### PR TITLE
#2322 - IER12 E2E Tests Additional Scenarios

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created two more scenarios from the previous PR (#2461) to-do list.
  - Has full-time disbursement feedback errors (DISE).
  - Has a disbursement that withheld funding due to a restriction (DISW).
  
- The first E2E above is invoking the processor without specifying the date as per [previous PR comment](https://github.com/bcgov/SIMS/pull/2461#discussion_r1379017653).
- New E2E factory created `createDisbursementFeedbackError`.
- Added an exception in case `getUploadedFile` was invoked but it was never called. The tests were failing with a not-so-clear error.
- Fixed an error in the IER12 query `BETWEEN` operator and removed a code TODO.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts [hunk 1]
```diff
@@ -390,4 +390,147 @@ describe(describeProcessorRootTest(QueueNames.IER12Integration), () => {
...
+    };
+    const application = await saveIER12TestInputData(db, testInputData, {
+      programYearPrefix: sharedProgramYearPrefix,
+      submittedDate: referenceSubmissionDate,
+    });
+    // Create a feedback error associated with the disbursement.
+    const [errorCode] = FULL_TIME_DISBURSEMENT_FEEDBACK_ERRORS;
+    const [disbursementSchedule] =
+      application.currentAssessment.disbursementSchedules;
+    const feedbackError = createDisbursementFeedbackError(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts:432]: Minor. Should we follow the naming convention of createFake*?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts [hunk 2]
```diff
@@ -390,4 +390,147 @@ describe(describeProcessorRootTest(QueueNames.IER12Integration), () => {
...
+      { disbursementSchedule },
+      { initialValues: { errorCode, updatedAt: dateUtils.addDays(-1) } },
+    );
+    await db.disbursementFeedbackErrors.save(feedbackError);
+
+    // Queued job.
+    // No date provided as it is expected that the disbursement feedback error
+    // date would ensure the IER12 record is generated.
+    const job = createIER12SchedulerJobMock();
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts:441]: I believe the IER here is running based on current date -1 and picking the IER record based on feedback error table.
Amazing to see this level of validation in E2E ❤️

File: sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts [hunk 3]
```diff
@@ -50,6 +50,11 @@ export function getUploadedFile(
   sshClientMock: DeepMocked<Client>,
 ): UploadedFile {
   const uploadedFile = {} as UploadedFile;
+  if (!sshClientMock.put.mock.calls.length) {
+    throw new Error(
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts:54]: 👍

File: sources/packages/backend/libs/utilities/src/date-utils.ts [hunk 4]
```diff
@@ -201,17 +201,17 @@ export function addToDateOnlyString(
...
-  return Between(
-    new Date(date.setHours(0, 0, 0, 0)),
-    new Date(date.setDate(date.getDate() + 1)),
-  );
-};
+export function dateEqualTo(date: Date): FindOperator<Date> {
+  const startDate = new Date(date);
+  startDate.setHours(0, 0, 0, 0);
+  const endDate = addDays(1, startDate);
+  return And(MoreThanOrEqual(startDate), LessThan(endDate));
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/utilities/src/date-utils.ts:213]: 👍

**Review by dheepak-aot — APPROVED** [review]
Great to see the level of validation for IER12. Approving with one very minor comment.

I believe there is going to be an upcoming PR for multiple institutions scenario right?

**Comment by andrewsignori-aot** [comment]
> Great to see the level of validation for IER12. Approving with one very minor comment.
> 
> I believe there is going to be an upcoming PR for multiple institutions scenario right?

Yes, the idea is still the same: try to finish what we can inside the sprint. I will take this scenario as a priority for the upcoming PR 😉

File: sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts [hunk 5]
```diff
@@ -50,6 +50,11 @@ export function getUploadedFile(
   sshClientMock: DeepMocked<Client>,
 ): UploadedFile {
```
Comment (sh16011993) [inline sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts:51]: Not a part of this PR but maybe we can update the comment to: "Get the parameters provided to the put method of the SSH client that represents **the remote file path and** the data that would be uploaded to the SFTP in a real scenario."

**Review by sh16011993 — APPROVED** [review]
Nice work 👍 @andrewsignori-aot

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts [hunk 6]
```diff
@@ -217,10 +220,8 @@ describe(describeProcessorRootTest(QueueNames.IER12Integration), () => {
 
     // Act
     const ier12Results = await processor.processIER12File(job);
-    db.application.createQueryBuilder();
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts:220]: 👍

**Review by andrepestana-aot — APPROVED** [review]
Looks good!

File: sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts:51]: Changed.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/institution-integration/ier12-integration/_tests_/e2e/ier12-integration.scheduler.e2e-spec.ts]: Changed.

**Review by guru-aot — COMMENTED** [review]
LGTM

**Pull request merged**